Modules/posixmodule.c

```diff
@@ -5088,384 +5088,392 @@ os__path_splitroot_impl(PyObject *module, path_t *path)
-                    result = 0;
+    DWORD attributes, reparseTag;
+    if ((testedType != PY_IFREG && testedType != PY_IFDIR)) {
+        FILE_ATTRIBUTE_TAG_INFO info;
+        if (!GetFileInformationByHandleEx(hfile, FileAttributeTagInfo, &info,
+                                          sizeof(info)))
+        {
+            return FALSE;
         }
+        attributes = info.FileAttributes;
+        reparseTag = info.ReparseTag;
     }
-    if (slow_path) {
-        if (_path.fd != -1) {
-            hfile = _Py_get_osfhandle_noraise(_path.fd);
-            close_file = FALSE;
-        }
-        else {
-            hfile = CreateFileW(_path.wide, FILE_READ_ATTRIBUTES, 0, NULL,
-                                OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
+    else {
+        FILE_BASIC_INFO info;
+        if (!GetFileInformationByHandleEx(hfile, FileBasicInfo, &info,
+                                          sizeof(info)))
+        {
+            return FALSE;
         }
-        if (hfile != INVALID_HANDLE_VALUE) {
-            if (GetFileInformationByHandleEx(hfile, FileBasicInfo, &info,
-                                            sizeof(info)))
-            {
-                result = info.FileAttributes & FILE_ATTRIBUTE_DIRECTORY;
+        attributes = info.FileAttributes;
+        reparseTag = 0;
+    }
+
+    if (attributes & FILE_ATTRIBUTE_REPARSE_POINT) {
+        return _testReparseTag(reparseTag, testedType);
+    }
+    else if (testedType == PY_IFREG) {
+        return ((fileDevType == FILE_TYPE_DISK) &&
+                !(attributes & FILE_ATTRIBUTE_DIRECTORY));
+    }
+    else if (testedType == PY_IFDIR) {
+        return attributes & FILE_ATTRIBUTE_DIRECTORY;
+    }
+
+    return FALSE;
+}
+
+static BOOL
+_testFileTypeByName(LPCWSTR path, int testedType)
+{
+    assert(testedType == PY_IFREG || testedType == PY_IFDIR ||
+           testedType == PY_IFLNK || testedType == PY_IFMNT ||
+           testedType == PY_IFLRP || testedType == PY_IFRRP);
+
+    FILE_STAT_BASIC_INFORMATION info;
+    if (_Py_GetFileInformationByName(path, FileStatBasicByNameInfo, &info,
+                                     sizeof(info)))
+    {
+        if (info.FileAttributes & FILE_ATTRIBUTE_REPARSE_POINT) {
+            if (testedType == PY_IFREG) {
+                if (info.FileAttributes & FILE_ATTRIBUTE_DIRECTORY) {
+                    return FALSE;
+                }
             }
-            else {
-                result = 0;
+            else if (testedType == PY_IFDIR) {
+                if (!(info.FileAttributes & FILE_ATTRIBUTE_DIRECTORY)) {
+                    return FALSE;
+                }
             }
-            if (close_file) {
-                CloseHandle(hfile);
+            else {
+                return _testReparseTag(info.ReparseTag, testedType);
             }
         }
+        else if (testedType == PY_IFREG) {
+            return !(info.FileAttributes & FILE_ATTRIBUTE_DIRECTORY);
```

Comment:
Queried `FileAttributes` are required to never be 0 for a filesystem file. The value must be `FILE_ATTRIBUTE_NORMAL` (0x0080) if no other attribute is set. On the other hand, for any kernel File object that doesn't reference a file or directory in a filesystem, I'd expect `FileAttributes` to be 0. I don't actually expect `FileStatBasicByNameInfo` to be implemented for anything but filesystem items, but just in case, let's ensure that `FileAttributes` is also non-zero here:

```suggestion
            return (info.FileAttributes && !(info.FileAttributes &
                    FILE_ATTRIBUTE_DIRECTORY));
```